README.md
# Lookup-Folder 0.7.0
Search a folder for files containing multiple strings.

The search strings are assumed to be URL encoded. There is an option to indicate
that the files are encoded as quoted printable (i.e. SMTP email files) and should be decoded.

This script is designed to be as fast as possible over WAN network shares.

Be sure to put final / if you are referring to folder rather than a file.
That is to say `C:/Windows/` and not `C:/Windows`. 

Note that you must use forward slashes even on Windows due to the way the `glob` function works.
```
perl Lookup-Folder.pl --folder //window.server/D$/mailroot/Pickup/* --search hello
```

Should work on Linux also.

### Example 1
```
perl Lookup-Folder.pl --search print --search user1 --folder ./

Search base path  : ./
Search target for : print
Search target for : user1
Max file age mins : none
Flags             :

Built file list in 0.005 seconds

[1] (age: 14.2 mins) critic.txt
    print ...  not found

[2] (age: 24.2 mins) cust1.eml
    print ...  not found

[3] (age: 184.2 mins) cust.eml
    print ...  not found

[4] (age: 1221.2 mins) 180_qponly.eml
    print ...  not found

[5] (age: 1224.2 mins) 180.eml
    print ...  FOUND
    user1 ...  not found

[6] (age: 4093.2 mins) LICENSE
    print ...  FOUND
    user1 ...  not found

[7] (age: 4093.2 mins) CHANGES.md
    print ...  FOUND
    user1 ...  not found

[8] (age: 4093.2 mins) Lookup-Folder.pl
    print ...  FOUND
    user1 ...  FOUND
    Success 'Lookup-Folder.pl' contains all search criteria!

[9] (age: 4093.2 mins) README.md
    print ...  FOUND
    user1 ...  FOUND
    Success 'README.md' contains all search criteria!

Searched files in 0.002 seconds

Found 2 matching files out of 9 files searched

```

### Example 2 - decode quoted printable files

```
perl Lookup-Folder.pl --search forgotten --search customer --folder ./*.eml --decode

Search base path  : ./*.eml
Search target for : forgotten
Search target for : customer
Max file age mins : none
Flags             : [decode quoted printable]

Built file list in 0.005 seconds

[1] (age: 25.9 mins) cust1.eml
    forgotten ...  FOUND
    customer ...  FOUND
    Success 'cust1.eml' contains all search criteria!

[2] (age: 185.9 mins) cust.eml
    forgotten ...  not found

[3] (age: 1222.9 mins) 180_qponly.eml
    forgotten ...  FOUND
    customer ...  FOUND
    Success '180_qponly.eml' contains all search criteria!

[4] (age: 1225.9 mins) 180.eml
    forgotten ...  FOUND
    customer ...  FOUND
    Success '180.eml' contains all search criteria!

Searched files in 0.001 seconds

Found 3 matching files out of 4 files searched
```

### Example 3 - search strings are URL encoded

```
perl Lookup-Folder.pl --search reset%20your%20password --search customer --folder ./180.eml --decode

Search base path  : ./180.eml
Search target for : reset your password
Search target for : customer
Max file age mins : none
Flags             : [decode quoted printable]

Built file list in 0.005 seconds

[1] (age: 1226.6 mins) 180.eml
    reset your password ...  FOUND
    customer ...  FOUND
    Success '180.eml' contains all search criteria!

Searched files in 0 seconds

Found 1 matching files out of 1 files searched
```

### Example 4 - stop after first matching file found

```
Lookup-Folder.pl --search reset%20your%20password --search customer --folder ./*.eml --decode --stop

Search base path  : ./*.eml
Search target for : reset your password
Search target for : customer
Max file age mins : none
Flags             : [stop] [decode quoted printable]

Built file list in 0.005 seconds

[1] (age: 27.3 mins) cust1.eml
    reset your password ...  not found

[2] (age: 187.3 mins) cust.eml
    reset your password ...  not found

[3] (age: 1224.3 mins) 180_qponly.eml
    reset your password ...  FOUND
    customer ...  FOUND
    Success '180_qponly.eml' contains all search criteria!

Searched files in 0 seconds

Found 1 matching files out of 3 files searched
```

### Example 5 - stop if files too old

```
Lookup-Folder.pl --search reset%20your%20password --search customer --folder ./*.eml --decode --stop --max_age 1200

Search base path  : ./*.eml
Search target for : reset your password
Search target for : customer
Max file age mins : 1200
Flags             : [stop] [decode quoted printable]

Built file list in 0.005 seconds

[1] (age: 28.9 mins) cust1.eml
    reset your password ...  not found

[2] (age: 188.9 mins) cust.eml
    reset your password ...  not found

[3] (age: 1225.9 mins) 180_qponly.eml - TOO OLD - REMAINGING FILES THIS OLD OR OLDER, STOPPING...

Searched files in 0 seconds

Found 0 matching files out of 3 files searched
```

CHANGES.md
# Lookup-Folder change log

https://github.com/Qarj/LookupFolder

---------------------------------------------------

## Release History:

### Version 0.7.0 - Mar 27, 2020
* Removed dependency on Windows, should now work on Linux also

### Version 0.06 - Mar 12, 2016
* major bug fixed - only worked for current directory - oops!!

### Version 0.05 - Mar 12, 2016
* examine file age and only search files younger than a certain age

### Version 0.04 - Mar 11, 2016
* stop after first matching file found

### Version 0.03 - Mar 11, 2016
* search strings are assumed to be URL encoded - to support spaces and other special characters

### Version 0.02 - Mar 11, 2016
* option to decoded quoted printable files implemented

### Version 0.01 - Mar 10, 2016
* search files in a folder for multiple criteria


Lookup-Folder.pl
#!/usr/bin/perl

# $Id$
# $Revision$
# $Date$

# perl Lookup-Folder.pl --folder ./test/*.txt --search hello --search WORLD
# perl Lookup-Folder.pl --folder ./test/*.txt --search hello --search WORLD --max_age 15
# perl Lookup-Folder.pl --folder ./test/*.txt --search hello --search WORLD --stop
# perl Lookup-Folder.pl --folder //window.server/D$/mailroot/Pickup/* --search hello
# 
# Note the direction of the slashes in the last example.

use strict;
use warnings;
use vars qw/ $VERSION /;

$VERSION = '0.7.0';

use File::Basename;
use Time::HiRes 'time';
use Getopt::Long;
use MIME::QuotedPrint; # for decoding quoted-printable
use File::Slurp;
use URI::Escape; # to convert the URL encoded search string to normal
use File::Glob qw(bsd_glob);

my (@opt_search, $opt_folder, $opt_stop, $opt_max_age, $opt_decode, $opt_version, $opt_help);

get_options();

decode_search_strings();

my $file_matches = 0;
my $files_checked = 0;
my (@files_to_check, @files_creation_time);
my $current_time = time;
print "Current Time: $current_time\n";

show_search_options();

build_list_of_files_to_check();

search_all_files();

print "\nFound $file_matches matching files out of $files_checked files searched\n";

#
# end of script - subroutines follow
#

# Build a list of files to examine in an array
sub build_list_of_files_to_check {

    my $start_time = time;

    my @raw_files = glob($opt_folder);

    my $i = 0;
    foreach (@raw_files) {
        $files_creation_time[$i] = (stat ($_))[9];
        $files_to_check[$i] = $_;
        print "$files_creation_time[$i]:$files_to_check[$i]\n";
        $i++;
    }

    # Sort the files from newest to oldest
    # This means we can stop reading the contents of files once we hit a specified age
    my @idx = sort { $files_creation_time[$b] <=> $files_creation_time[$a] } 0 .. $#files_creation_time;
    @files_creation_time = @files_creation_time[@idx];
    @files_to_check = @files_to_check[@idx];

    my $run_time = (int(1000 * (time - $start_time)) / 1000);
    print "Built file list in $run_time seconds\n";

    return;
}

# Iterate over the files
sub search_all_files {

    my $start_time = time;

    for my $i (0 .. $#files_to_check)
    {
        $files_checked = $files_checked + 1;
        print "\n["."$files_checked".'] ';

        my $file_age_mins = sprintf '%.1f', ($current_time - $files_creation_time[$i]) / 60;
        print "(age: $file_age_mins mins) ";
        if (defined $opt_max_age) {
            if ($file_age_mins > ($opt_max_age)) {
                print "$files_to_check[$i] - TOO OLD - REMAINGING FILES THIS OLD OR OLDER, STOPPING...\n";
                last;
            }
        }
        print "$files_to_check[$i]\n";

        examine_file ($files_to_check[$i]);
        if (defined $opt_stop && ($file_matches > 0) ) {
            # option to stop after first match is enabled ...
            last; # ... so do not process any more files
        }
    }

    my $run_time = (int(1000 * (time - $start_time)) / 1000);
    print "\nSearched files in $run_time seconds\n";

    return;
}

# Examine one file
sub examine_file {
    my ($filename) = @_;

    my $raw_folder = dirname($opt_folder.'dummy');

    my $text = read_file($filename);

    if (defined $opt_decode) {
        $text = decode_qp($text); ## decode the response output assuming it was quoted printable (i.e. SMTP email format)
    }

    foreach my $search (@opt_search) {

        print "    $search ... ";

        my $match = 0;

        if ($text =~ m/$search/i) {

            # search string has been found in this file
            $match = 1;

            print " FOUND\n";
        }

        if ($match) {
            # great news
        } else {
            print " not found\n";
            return; # this search string was not found in this file, so it is a fail
        }
    }

    # if we made it to here, it means all search strings were found in this file
    $file_matches = $file_matches + 1;

    print "    Success '$filename' contains all search criteria!\n";

    return;
}
#------------------------------------------------------------------
sub decode_search_strings {  # the supplied search strings are assumed to be URL encoded

    foreach (@opt_search) {
        $_ = uri_unescape($_);
    }

    return;
}

sub show_search_options {
    print "\n";
    print "Search base path  : $opt_folder\n";

    foreach (@opt_search) {
        print "Search target for : $_ \n";
    }

    print 'Max file age mins : ';
    if (defined $opt_max_age) {
        print "$opt_max_age\n";
    } else {
        print "none\n";
    }

    print 'Flags             :';
    if (defined $opt_stop) { print ' [stop]'; }
    if (defined $opt_decode) { print ' [decode quoted printable]'; }
    print "\n\n";

    return;
}

#------------------------------------------------------------------
sub get_options {  #shell options

    Getopt::Long::Configure('bundling');
    GetOptions(
        's|search=s'  => \@opt_search,
        'f|folder=s'   => \$opt_folder,
        'm|max_age=i'   => \$opt_max_age,
        't|stop'   => \$opt_stop,
        'q|decode'   => \$opt_decode,
        'v|V|version' => \$opt_version,
        'h|help'      => \$opt_help,
        )
        or do {
            print_usage();
            exit;
        };
    if ($opt_version) {
        print_version();
        exit;
    }

    if ($opt_help) {
        print_version();
        print_usage();
        exit;
    }

    if (not defined $opt_search[0]) {
        print "\nERROR: Search string[s] must be specified\n\n";
        print_usage();
        exit;
    }

    if (not defined $opt_folder) {
        print "\nERROR: Target folder must be specified\n\n";
        print_usage();
        exit;
    }

    return;
}

sub print_version {
    print "\nLookupFolder version $VERSION\nFor more info: https://github.com/Qarj/LookupFolder\n\n";
    return;
}

sub print_usage {
    print <<'EOB'

Usage: LookupFolder.pl <<options>>

-s|--search     SEARCH STRING, MANY OK    --search user1 --search my%20name
-f|--folder     TARGET FOLDER TO SEARCH   --folder \\IRON\D$\email\pickup\*.eml
-m|--max_age    MAX AGE OF FILES - MINS   --max_age 15
-t|--stop       STOP AFTER FIRST MATCH    --stop
-q|--decode     DECODE QUOTED PRINTABLE   --decode
-v|--version                              -v
-h|--help                                 -h

or

LookupFolder.pl -v|--version
LookupFolder.pl -h|--help
EOB
;
return;
}
#------------------------------------------------------------------


LookupFolder.pl
#!/usr/bin/perl

# $Id$
# $Revision$
# $Date$

use strict;
use warnings;
use vars qw/ $VERSION /;

$VERSION = '0.05';

use File::Basename;
use Time::HiRes 'time';
use Time::Piece;
use Getopt::Long;
use MIME::QuotedPrint; # for decoding quoted-printable
use File::Slurp;
use URI::Escape; # to convert the URL encoded search string to normal

my (@opt_search, $opt_folder, $opt_stop, $opt_max_age, $opt_decode, $opt_version, $opt_help);

get_options();

decode_search_strings();

#Example:
#LookupFolder.pl --search forgotten --search customer --folder .\*.eml

#To Implement:
# - determine date format - use: echo. | date

my $file_matches = 0;
my $files_checked = 0;
my (@files_to_check, @files_creation_time);
my $current_time = Time::Piece->new;

show_search_options();

build_list_of_files_to_check();

search_all_files();

print "\nFound $file_matches matching files out of $files_checked files searched\n";

#
# end of script - subroutines follow
#

# Build a list of files to examine in an array
sub build_list_of_files_to_check {

    my $start_time = time;

    #dir folder /A-D /N /O-D /TC /-C
    # /A-D  do not display directories
    # /N    use new long list format with filenames on far right
    # /O-D  sort by date/time with youngest first
    # /TC   sort date/time is creation
    # /-C   disable display of thousands separator for file size
    # /4    display four-digit years
    my @raw_files = (`dir /A-D /N /O-D /TC /-C /4 "$opt_folder"`);

    #foreach (@raw_files) {
    #    print "$_\n";
    #}

    my $i = 0;
    foreach (@raw_files) {
        # regex matches date, time, file size then anything after that is the file name
        if ($_ =~ m{^([\d/]+[ ]+[\d:]+)[ ]+[\d]+[ ](.+)}i ) {
            $files_creation_time[$i] = Time::Piece->strptime($1, '%d/%m/%Y  %H:%M'); # european date format assumed
            $files_to_check[$i] = $2;
            #print "$files_creation_time[$i]:$files_to_check[$i]\n";
            $i++;
        }
    }

    # debug - check the file names captured
    #for my $j (0 .. $#files_to_check) {
    #    print "$files_creation_time[$j] ";
    #    print "$files_to_check[$j]\n";
    #}
    #print "\n";

    my $run_time = (int(1000 * (time - $start_time)) / 1000);
    print "Built file list in $run_time seconds\n";

    return;
}

# Iterate over the files
sub search_all_files {

    my $start_time = time;

    for my $i (0 .. $#files_to_check)
    {
        $files_checked = $files_checked + 1;
        print "\n["."$files_checked".'] ';

        my $file_age_mins = sprintf '%.1f', ($current_time - $files_creation_time[$i]) / 60;
        print "(age: $file_age_mins mins) ";
        if (defined $opt_max_age) {
            if ($file_age_mins > ($opt_max_age)) {
                print "$files_to_check[$i] - TOO OLD - REMAINGING FILES THIS OLD OR OLDER, STOPPING...\n";
                last;
            }
        }
        print "$files_to_check[$i]\n";

        examine_file ($files_to_check[$i]);
        if (defined $opt_stop && ($file_matches > 0) ) {
            # option to stop after first match is enabled ...
            last; # ... so we hot foot it outa here
        }
    }

    my $run_time = (int(1000 * (time - $start_time)) / 1000);
    print "\nSearched files in $run_time seconds\n";

    return;
}

# Examine one file
sub examine_file {
    my ($filename) = @_;

    chomp $filename; # remove trailing \n

    # swap back slash to forward slash
    my $linux_file_name = $filename;
    $linux_file_name =~ s{\\}{/}g;

    my $text = read_file($filename);

    if (defined $opt_decode) {
        $text = decode_qp($text); ## decode the response output
    }

    foreach my $search (@opt_search) {

        print "    $search ... ";

        my $match = 0;

        if ($text =~ m/$search/i) {

            # search string has been found in this file
            $match = 1;

            print " FOUND\n";
        }

        if ($match) {
            # great news
        } else {
            print " not found\n";
            return; # this search string was not found in this file, so it is a fail
        }
    }

    # if we made it to here, it means all search strings were found in this file
    $file_matches = $file_matches + 1;

    print "    Success '$filename' contains all search criteria!\n";

    return;
}
#------------------------------------------------------------------
sub decode_search_strings {  # the supplied search strings are assumed to be URL encoded

    foreach (@opt_search) {
        $_ = uri_unescape($_);
    }

    return;
}

sub show_search_options {
    print "\n";
    print "Search base path  : $opt_folder\n";

    foreach (@opt_search) {
        print "Search target for : $_ \n";
    }

    print 'Max file age mins : ';
    if (defined $opt_max_age) {
        print "$opt_max_age\n";
    } else {
        print "none\n";
    }

    print 'Flags             :';
    if (defined $opt_stop) { print ' [stop]'; }
    if (defined $opt_decode) { print ' [decode quoted printable]'; }
    print "\n\n";

    return;
}

#------------------------------------------------------------------
sub get_options {  #shell options

    Getopt::Long::Configure('bundling');
    GetOptions(
        's|search=s'  => \@opt_search,
        'f|folder=s'   => \$opt_folder,
        'm|max_age=i'   => \$opt_max_age,
        't|stop'   => \$opt_stop,
        'q|decode'   => \$opt_decode,
        'v|V|version' => \$opt_version,
        'h|help'      => \$opt_help,
        )
        or do {
            print_usage();
            exit;
        };
    if ($opt_version) {
        print_version();
        exit;
    }

    if ($opt_help) {
        print_version();
        print_usage();
        exit;
    }

    if (not defined $opt_search[0]) {
        print "\nERROR: Search string[s] must be specified\n\n";
        print_usage();
        exit;
    }

    if (not defined $opt_folder) {
        print "\nERROR: Target folder must be specified\n\n";
        print_usage();
        exit;
    }

    return;
}

sub print_version {
    print "\nLookupFolder version $VERSION\nFor more info: https://github.com/Qarj/LookupFolder\n\n";
    return;
}

sub print_usage {
    print <<'EOB'

Usage: LookupFolder.pl <<options>>

-s|--search     SEARCH STRING, MANY OK    --search user1 --search my%20name
-f|--folder     TARGET FOLDER TO SEARCH   --folder \\IRON\D$\email\pickup\*.eml
-m|--max_age    MAX AGE OF FILES - MINS   --max_age 15
-t|--stop       STOP AFTER FIRST MATCH    --stop
-q|--decode     DECODE QUOTED PRINTABLE   --decode
-v|--version                              -v
-h|--help                                 -h

or

LookupFolder.pl -v|--version
LookupFolder.pl -h|--help
EOB
;
return;
}
#------------------------------------------------------------------
